### PR TITLE
Extracting cross-model uniqueness validator

### PR DESCRIPTION
**Pull request by jeremyf**
## What type of PR is this? (check all applicable)

- [x] Refactor

## Description

Prior to this commit, we had somewhat duplicated logic across four
models.  In adding this validator we're consolidating the logic and
tidying up each of the other models; letting the validator know which
models are part of the "uniqueness gang."

I made the decision to include the test for `sitemap-` as this was
partially applied to two of the models.

## Related Tickets & Documents

None.

## QA Instructions, Screenshots, Recordings

None, though please read through my logic.

### UI accessibility concerns?

None.

## Added/updated tests?

- [x] Yes

## [Forem core team only] How will this change be communicated?

- [x] I will share this change internally with the appropriate teams

### CODE REVIEW

File: app/validators/unique_cross_model_slug_validator.rb [hunk 1]
```diff
@@ -0,0 +1,45 @@
+##
+# Validates if the give attribute is used across the reserved spaces.
+class UniqueCrossModelSlugValidator < ActiveModel::EachValidator
+  class_attribute :model_and_attribute_name_for_uniquness_test
```
Comment (citizen428) [inline app/validators/unique_cross_model_slug_validator.rb:4]:
```suggestion
  class_attribute :model_and_attribute_name_for_uniqueness_test
```

File: app/validators/unique_cross_model_slug_validator.rb [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+  # Why a class attribute?  Allow for other implementations to extend
+  # this behavior.
+  self.model_and_attribute_name_for_uniquness_test = {
```
Comment (citizen428) [inline app/validators/unique_cross_model_slug_validator.rb:8]:
```suggestion
  self.model_and_attribute_name_for_uniqueness_test = {
```

File: app/validators/unique_cross_model_slug_validator.rb [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+  #
+  # @return [TrueClass] if the value already exists across the various classes.
+  # @return [FalseClass] if the given value is not already used.
+  #
+  # @see CLASS_AND_ATTRIBUTE_NAME_FOR_UNIQUNESS_TEST
+  def already_exists?(value:, record:)
+    return false unless value
+    return true if value.include?("sitemap-")
+
+    model_and_attribute_name_for_uniquness_test.detect do |model, attribute|
```
Comment (citizen428) [inline app/validators/unique_cross_model_slug_validator.rb:39]:
```suggestion
    model_and_attribute_name_for_uniqueness_test.detect do |model, attribute|
```

File: app/validators/unique_cross_model_slug_validator.rb [hunk 4]
```diff
@@ -0,0 +1,45 @@
+##
+# Validates if the give attribute is used across the reserved spaces.
+class UniqueCrossModelSlugValidator < ActiveModel::EachValidator
+  class_attribute :model_and_attribute_name_for_uniquness_test
+
+  # Why a class attribute?  Allow for other implementations to extend
```
Comment (citizen428) [inline app/validators/unique_cross_model_slug_validator.rb:6]: Could you elaborate on the possible extensions you have in mind and why a regular Ruby method that subclasses can override wasn't an option for that? It seems a bit unlikely to me that someone would make use of the `UniqueCrossModelSlugValidator.model_and_attribute_name_for_uniqueness_test = ... ` accessor `class_attribute` provides as people don't usually interact with validators so directly.

**Review by citizen428 — CHANGES_REQUESTED** [review]
Requesting changes so we fix the typo before merging, the other comment is just asking for clarification.

File: app/validators/unique_cross_model_slug_validator.rb (hunk 4 above)
Comment (jeremyf) [reply app/validators/unique_cross_model_slug_validator.rb:6]: I think I got ahead of myself in preliminary refactoring (this started as a constant).  But I'll go into my thoughts.

I'm coming from a past community that released their application as a Rails Engine.  You would basically mount the engine at "/", so the code was not visible in the application.  This meant we needed to expose more configuration options and eschewed the idea of overriding classes.  We tried to refactor/build that engine with interfaces in mind.

I'm uncertain how others are forking the application and making their modifications (and keeping up to date with our modifications).

The class attribute allows an implementor to add a line to their application initializer to append a Model / Attribute pair to the existing system (see below):

```ruby
UniqueCrossModelSlugValidator.model_and_attribute_name_for_uniqueness_test[NewModel] = :slug
```

Were they to overwrite the method (by going into the validator) then they would have potential future merge conflicts to address.

I suppose this is my baby-step into the larger architectural question of "How do we envision folks extending the Forem application?"

Comment (citizen428) [reply app/validators/unique_cross_model_slug_validator.rb:6]: Thanks for the context, @jeremyf. Personally, I would prefer if we make decisions that go in the direction of overall app architecture in a more collaborative way, i.e. via a forem.team discussion. In the meantime, a constant wrapped in a method would be more in line with what we have elsewhere.

Comment (jeremyf) [reply app/validators/unique_cross_model_slug_validator.rb:6]: I made the change, and definitely will go the route of discussions.

**Review by citizen428 — APPROVED** [review]

**Review by Ridhwana — APPROVED** [review]

File: app/validators/unique_cross_model_slug_validator.rb (hunk 4 above)
Comment (citizen428) [reply app/validators/unique_cross_model_slug_validator.rb:6]: Sorry, when you said "made the change" I thought you changed the code back to using a constant. Apparently you didn't, so I'm not quite sure what you actually meant/if I misinterpreted your statement?

**Pull request merged**